### ISSUE TITLE
Docker Windows CE, DNS not working from docker-compose up

### ISSUE THREAD

**Issue by jasonterando** [docker/docker#31658]
**Description**

After upgrading from Docker for Windows 1.13.1 to CE, containers running in Docker Compose can no longer resolve names via DNS.  The same images launched via "docker run" do allow DNS name resolution.  

My Docker compose file has no custom networking or DNS settings.  

When running in either version of Docker for Windows, /etc/resolv.conf shows the nameserver as 192.168.65.1 when the container is launched via "docker run", and 127.0.0.11 when launched via "docker-compose".  

**Steps to reproduce the issue:**
1.  Install Docker for Windows 17.03.0-ce
2.  In an empty directory, create a Dockerfile that is just "FROM httpd"
3.  In the same directory, create a Docker Compose file like this:

```
version: '3'

services:
  httpd:
    build: 
       context: .
       dockerfile: dockerfile.test

````

4.  Execute "docker-compose build"
5.  Execute "docker run httpd"
6.  Execute "docker ps" to get the container name
7.  Execute "docker exec -it CONTAINER_NAME bash"
8.  In the shell, do a "ping www.google.com", it works
9.  Exit the shell, run "docker kill CONTAINER_NAME" 
10. Execute "docker-compose up"
11. Repeat steps 6-8, and in my case, I could not reach www.google.com
12. In the shell, confirm you can ping Google by IP address (74.125.203.106)
13. Exit the shell, run "docker kill CONTAINER_NAME"
14. If you want, uninstall Docker CE, reinstall 1.13.1, and execute 10-13, you should now be able to ping by DNS name

**Describe the results you received:**

When running containers via Docker Compose using the most recent release of Docker for Windows CE,  I am unable to resolve host name via DNS

**Describe the results you expected:**

When running containers via Docker Compose using the most recent release of Docker for Windows CE,  I am would like to resolve host name via DNS, just like I could in 1.13.1

**Additional information you deem important (e.g. issue happens only occasionally):**

Nothing very special about my machine setup.  Windows 10 Professional, Hyper-V is healthy and defaults have never been touch.  

I have tried CE stable and edge, same results.

I have built images that include nslookup, telnet, etc. and it looks like I am not able to make port 53 connections to the either "internal" 127.0.0.11 DNS or external (8.8.8.8). 

**Output of `docker version`:**

```
Docker version 17.03.0-ce, build 60ccb22
```

**Output of `docker info`:**

```
Containers: 31
 Running: 0
 Paused: 0
 Stopped: 31
Images: 574
Server Version: 17.03.0-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host ipvlan macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 977c511eda0925a723debdc94d09459af49d082a
runc version: a01dafd48bc1c7cc12bdb01206f9fea7dd6feb70
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.12-moby
Operating System: Alpine Linux v3.5
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 1.941 GiB
Name: moby
ID: LXV7:I3LF:OGTM:HP6P:QUT3:5IQ4:H6YE:GMDE:FDM3:H4YK:2LEP:4HPO
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 13
 Goroutines: 22
 System Time: 2017-03-08T17:32:44.5063104Z
 EventsListeners: 0
Registry: https://index.docker.io/v1/
Experimental: true
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Plenty of resources, stuff was fine under 13.1.  Windows 10 Professional 64 bit with 16 GB of RAM, i7-5557U CPU.

**Comment by thaJeztah** [comment]
I'm not able to reproduce this on Docker for Mac 17.03.0-ce. What DNS servers do you have configured on your host?

ping @rneugeba any issues known on Docker for Windows?

**Comment by jasonterando** [comment]
Thanks for the reply.

I have four DNS servers on my Ethernet adapter that are returned via DHCP, 10.2.10.8, 10.31.1.17, 10.31.1.141, 10.31.1.36 and 8.8.8.8.  For the Hytper-V Virtual Ethernet Adapter, there are three IPV6 DNS servers listed, fec0:0:0:ffff::1%1, fec0:0:0:ffff::2%1 and fec0:0:0:ffff::3%1.  

Also, Virtual Switch Manager for Hyper-V is set to Internal Network (which I thought might have been the issue, but switching it to External Network had no effect, and Internal Network works okay on 1.13.1

**Comment by jasonterando** [comment]
For giggles, here's Docker info from 1.13.1

```
Containers: 31
 Running: 0
 Paused: 0
 Stopped: 31
Images: 574
Server Version: 1.13.1
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host ipvlan macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: aa8187dbd3b7ad67d8e5e3a15115d3eef43a7ed1
runc version: 9df8b306d01f59d3a8029be411de015b7304dd8f
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.8-moby
Operating System: Alpine Linux v3.5
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 1.941 GiB
Name: moby
ID: L34O:GPC6:A3NO:2UL5:H7UJ:DA5G:BWNY:WTQM:55T3:KW2G:4MLB:ZUF3
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 13
 Goroutines: 22
 System Time: 2017-03-08T19:13:40.9568337Z
 EventsListeners: 0
Registry: https://index.docker.io/v1/
Experimental: true
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Comment by thaJeztah** [comment]
What happens if you manually override the dns server for the docker daemon, to use `8.8.8.8`? https://docs.docker.com/docker-for-windows/#network

**Comment by jasonterando** [comment]
So, I got some unexpected results, not sure if these are good or bad...

1. Setting DNS in the Docker setting screen to 8.8.8.8 caused DNS lookups in the container launched with "docker run" to fail (as well as the same image being run with docker compose)
2. I checked /etc/resolv.conf and both were the same values they were before (i.e. I did not see "nameserver 8.8.8.8"), but I am a relative Docker newb, so perhaps that really isn't a thing.
3. Resolving DNS back to Automatic got me to the original behavior, image launched via run could resolve names, image launched via compose up could not

**Comment by thaJeztah** [comment]
Ok, so the difference between starting through `docker-compose` and `docker run` is that Docker compose creates a custom network for the stack, and attaches the container to it. To see if that's the difference, can you try creating a network, and running a container on that network?

```bash
docker network create foo
docker run --network=foo --rm httpd ping -c1 www.google.com
```

If that's exposing the same problem as docker compose, that may be the difference (I don't have a solution yet, may have to try getting other people on this issue from the Docker for Windows team, but having the info ready for them to look at may be helpful :smile:)

**Comment by rneugeba** [comment]
@djs55 is probably more knowledgable re DNS lookups in Docker for Windows

**Comment by jasonterando** [comment]
I think you nailed it, or least have a heck of a lead on it.  In the excerpt below I run the ping with the "default" bridge (it works); then create a "foo" network and run the ping (it fails) and then run the ping with "default" bridge again (it works).  

```
PS C:\projects\prism-docker\test> docker run --rm httpd ping -c1 www.google.com
PING www.google.com (74.125.204.104): 56 data bytes
64 bytes from 74.125.204.104: icmp_seq=0 ttl=37 time=0.268 ms
--- www.google.com ping statistics ---
1 packets transmitted, 1 packets received, 0% packet loss
round-trip min/avg/max/stddev = 0.268/0.268/0.268/0.000 ms
PS C:\projects\prism-docker\test> docker network create foo
7d6b2c2ffa521eb81486164e61839218646f4d83e8e01963e0e93bc0018f005a
PS C:\projects\prism-docker\test> docker run --network=foo --rm httpd ping -c1 www.google.com
ping: unknown host
PS C:\projects\prism-docker\test> docker run --rm httpd ping -c1 www.google.com
PING www.google.com (74.125.204.104): 56 data bytes
64 bytes from 74.125.204.104: icmp_seq=0 ttl=37 time=0.873 ms
--- www.google.com ping statistics ---
1 packets transmitted, 1 packets received, 0% packet loss
round-trip min/avg/max/stddev = 0.873/0.873/0.873/0.000 ms
```

I ran an inspect on both networks.  It looks like there are some Options under the "Bridge" network which may be relevant.  I wonder if there are some default options that may have changed.

```
PS C:\projects\prism-docker\test> docker network inspect bridge
[
    {
        "Name": "bridge",
        "Id": "8e5e745538ac220c01f70ba656626be23147a3ace3096ec117fcb16712b4eef0",
        "Created": "2017-03-08T22:46:33.3993197Z",
        "Scope": "local",
        "Driver": "bridge",
        "EnableIPv6": false,
        "IPAM": {
            "Driver": "default",
            "Options": null,
            "Config": [
                {
                    "Subnet": "172.17.0.0/16",
                    "Gateway": "172.17.0.1"
                }
            ]
        },
        "Internal": false,
        "Attachable": false,
        "Containers": {},
        "Options": {
            "com.docker.network.bridge.default_bridge": "true",
            "com.docker.network.bridge.enable_icc": "true",
            "com.docker.network.bridge.enable_ip_masquerade": "true",
            "com.docker.network.bridge.host_binding_ipv4": "0.0.0.0",
            "com.docker.network.bridge.name": "docker0",
            "com.docker.network.driver.mtu": "1500"
        },
        "Labels": {}
    }
]
PS C:\projects\prism-docker\test> docker network inspect foo
[
    {
        "Name": "foo",
        "Id": "7d6b2c2ffa521eb81486164e61839218646f4d83e8e01963e0e93bc0018f005a",
        "Created": "2017-03-08T22:50:31.601489Z",
        "Scope": "local",
        "Driver": "bridge",
        "EnableIPv6": false,
        "IPAM": {
            "Driver": "default",
            "Options": {},
            "Config": [
                {
                    "Subnet": "172.20.0.0/16",
                    "Gateway": "172.20.0.1"
                }
            ]
        },
        "Internal": false,
        "Attachable": false,
        "Containers": {},
        "Options": {},
        "Labels": {}
    }
]

```

Thanks again for the follow-up and patience working with a new user.

**Comment by thaJeztah** [comment]
ping @djs55 could you have a look?

**Comment by jasonterando** [comment]
Hi, anything I can do to help provide additional diagnostic info?  For now, staying with 1.13.1 works around my issue, but it doesn't seem like the best long-term strategy.

**Comment by djs55** [comment]
Sorry for the delay in responding.

Could you re-run your tests and then upload diagnostics using the Whale menu -> "Diagnose & Feedback" -> "Upload a diagnostic..."?The diagnostics should include logs and some DNS packet captures from inside the VM which should allow me to see what's happening.  I'd particularly like to compare and contrast

- a good DNS lookup via the bridge network
- a failing lookup via the "foo" network

As you've spotted, there are 2 DNS servers in Docker for Windows. On a default bridge network there is only `192.168.65.1`: this forwards from the VM to the outside world. When using the "foo" network there is also a server on `127.0.0.11` which helps resolve container names to internal IPs but still forwards other requests to `192.168.65.1`.

Thanks again for your report!

**Comment by jasonterando** [comment]
Hi, when running my tests, I got slightly different results, although still borked.
tl;dr - 1.13.1 allowed me to build and do an nslookup from within a container, 17.03.0 won't even let me do a build now.

Working 1.13.1 diagnostics:
C7ED5F5F-3273-4BDE-9A44-A018CA0CBD67/2017-03-16_11-33-32
Not working 17.03.0 diagnistics:
C7ED5F5F-3273-4BDE-9A44-A018CA0CBD67/2017-03-16_12-03-27

**Trial #1:  Installed Docker Windows 1.13.1, Factory Reset**
Diagnostics: C7ED5F5F-3273-4BDE-9A44-A018CA0CBD67/2017-03-16_11-33-32

```
root@31af0f13595a:/usr/local/apache2# nslookup --debug www.google.com
*** Invalid option: -debug
Server:         127.0.0.11
Address:        127.0.0.11#53

Non-authoritative answer:
Name:   www.google.com
Address: 64.233.188.105
Name:   www.google.com
Address: 64.233.188.106
Name:   www.google.com
Address: 64.233.188.99
Name:   www.google.com
Address: 64.233.188.103
Name:   www.google.com
Address: 64.233.188.147
Name:   www.google.com
Address: 64.233.188.104

root@31af0f13595a:/usr/local/apache2# nslookup --debug arl-dev-db.digitalroom.com
*** Invalid option: -debug
Server:         127.0.0.11
Address:        127.0.0.11#53

Non-authoritative answer:
Name:   arl-dev-db.digitalroom.com
Address: 10.52.111.223

```
**Trial #2: Installed Docker 17.03.0-ce build 60ccb2, Factory Reset**
Diagnostics: C7ED5F5F-3273-4BDE-9A44-A018CA0CBD67/2017-03-16_11-53-21
Could not do a build (same error as next trial, see below)

**Trial #3: Uninstalled Docker, removed DockerNAT, rebooted, reinstalled 17.03.0-c3**
Still cannot do a build, did a wget after the build which gave me the expected UNAUTHORIZED error, which shows that my machine can get to the Docker registry
Diagnostics: C7ED5F5F-3273-4BDE-9A44-A018CA0CBD67/2017-03-16_12-03-27

```
PS C:\projects\docker.test> docker-compose build --no-cache
Building httpd
Step 1/1 : FROM httpd
ERROR: Service 'httpd' failed to build: Get https://registry-1.docker.io/v2/: net/http: request canceled while waiting f
or connection (Client.Timeout exceeded while awaiting headers)
PS C:\projects\docker.test> wget https://registry-1.docker.io/v2/
wget : {"errors":[{"code":"UNAUTHORIZED","message":"authentication required","detail":null}]}
At line:1 char:1
+ wget https://registry-1.docker.io/v2/
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (System.Net.HttpWebRequest:HttpWebRequest) [Invoke-WebRequest], WebExc
   eption
    + FullyQualifiedErrorId : WebCmdletWebResponseException,Microsoft.PowerShell.Commands.InvokeWebRequestCommand
```

**Comment by jasonterando** [comment]
Hi - FYI I will be on vacation for a week starting on Friday.  If there is anything I can provide between now and then to help out, please let me know.  Thanks!

**Comment by jasonterando** [comment]
Newest version 17.03.1-ce-win5 (10743) makes everything Shiny & Chrome - just had to rebuild images.

**Comment by thaJeztah** [comment]
excellent, thanks @jasonterando !

**Comment by weslleycamilo** [comment]
I still facing it on docker-compose and by docker run itself.. 
i am using the stable version and it didn't work out to me.

Is there any clue why i cant ping any running container ? 

I can still ping www.google.com.br

P.S. I do have the same version running on  other servers and it is working fine.

**Comment by radekbaranowski** [comment]
This issue is still happening on Windows 10 and latest build of Docker:
dns resolution of internet domains eg. jenkins update) works with _docker run_ whereas fails when running the same container with _docker-compose_

**Comment by jasonterando** [comment]
@noraab  - I would open up a separate issue.  What you're describing is different than what I was experiencing.  If you don't get anywhere, maybe something is borked in Hyper-V.  You might try uninstalling Docker and Hyper-V, and then re-install both.
